File: lua/armstrong-numbers/armstrong-numbers.lua
local ArmstrongNumbers = {}

local function digits_len(number)
    if number == 0 then
        return 1
    end

    local digits = 0

    while number > 0 do
        digits = digits + 1
        number = math.floor(number / 10)
    end

    return digits
end

function ArmstrongNumbers.is_armstrong_number(number)
    local pow = digits_len(number)
    local sum = 0
    local x = number

    while x > 0 do
        local digit = x % 10
        sum = sum + (digit ^ pow)

        x = math.floor(x / 10)
    end

    return sum == number
end

return ArmstrongNumbers

File: c/armstrong-numbers/armstrong_numbers.c
#include <math.h>
#include "armstrong_numbers.h"

int num_of_digits(int candidate) {
    int digits = 0;

    while (candidate > 0) {
        digits++;
        candidate /= 10;
    }

    return digits;
}

bool is_armstrong_number(int candidate) {
    int digits = num_of_digits(candidate);
    int sum = 0;

    int x = candidate;
    for (int i = 0; i < digits; i++) {
        int digit = x % 10;
        sum += pow(digit, digits);
        x /= 10;
    }

    return sum == candidate;
}
